File: touch/src/main/java/de/nicidienase/chaosflix/touch/browse/BrowseActivity.java
package de.nicidienase.chaosflix.touch.browse;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.transition.Slide;
import android.transition.TransitionInflater;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;

import de.nicidienase.chaosflix.R;
import de.nicidienase.chaosflix.common.entities.recording.persistence.PersistentEvent;
import de.nicidienase.chaosflix.touch.OnEventSelectedListener;
import de.nicidienase.chaosflix.touch.activities.AboutActivity;
import de.nicidienase.chaosflix.touch.eventdetails.EventDetailsActivity;

public class BrowseActivity extends AppCompatActivity implements
		ConferencesTabBrowseFragment.OnConferenceListFragmentInteractionListener,
		EventsListFragment.OnEventsListFragmentInteractionListener,
		OnEventSelectedListener {

	private static final String TAG = BrowseActivity.class.getSimpleName();
	private Toolbar toolbar;
	private ActionBarDrawerToggle drawerToggle;
	private DrawerLayout drawerLayout;

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_browse);

		toolbar = findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);

		drawerLayout = findViewById(R.id.drawer_layout);
		drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.drawer_open, R.string.drawer_close);
		drawerLayout.addDrawerListener(drawerToggle);
		drawerToggle.syncState();

		resetToolbar();

		NavigationView navigationView = findViewById(R.id.navigation_view);
		navigationView.setNavigationItemSelectedListener(item -> {
			switch (item.getItemId()) {
				case R.id.nav_recordings:
					showConferencesFragment();
					break;
				case R.id.nav_bookmarks:
					showBookmarksFragment();
					break;
				case R.id.nav_inprogress:
					showInProgressFragment();
					break;
				case R.id.nav_about:
					showAboutPage();
					break;
				case R.id.nav_streams:
				case R.id.nav_preferences:
				default:
					Snackbar.make(drawerLayout, "Not implemented yet", Snackbar.LENGTH_SHORT).show();
					break;
			}
			drawerLayout.closeDrawers();
			return true;
		});

		if (savedInstanceState == null) {
			showConferencesFragment();
		}
	}

	@Override
	public void onPostCreate(@Nullable Bundle savedInstanceState, @Nullable PersistableBundle persistentState) {
		super.onPostCreate(savedInstanceState, persistentState);
		drawerToggle.syncState();
	}

	@Override
	public void onConfigurationChanged(Configuration newConfig) {
		super.onConfigurationChanged(newConfig);
		drawerToggle.onConfigurationChanged(newConfig);
	}

	//	@Override
//	public boolean onCreateOptionsMenu(Menu menu) {
//		getMenuInflater().inflate(R.menu.browse_menu, menu);
//		return true;
//	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (drawerToggle.onOptionsItemSelected(item)) {
			return true;
		}
		switch (item.getItemId()) {
			case android.R.id.home:
				drawerLayout.openDrawer(GravityCompat.START);
				return true;
			case R.id.action_show_bookmarks:
				showBookmarksFragment();
				return true;
			case R.id.action_show_inprogress:
				showInProgressFragment();
				return true;
			case R.id.action_about:
				showAboutPage();
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	private int getNumColumns() {
		return getResources().getInteger(R.integer.num_columns);
	}

	private int showConferencesFragment() {
		resetToolbar();
		return getSupportFragmentManager()
				.beginTransaction()
				.replace(R.id.fragment_container, ConferencesTabBrowseFragment.newInstance(getNumColumns()))
				.commit();
	}

	private void resetToolbar() {
		getSupportActionBar().setDisplayShowHomeEnabled(false);
		getSupportActionBar().setTitle(R.string.app_name);
//		getSupportActionBar().setLogo(R.drawable.icon_notext);
//		getSupportActionBar().setDisplayUseLogoEnabled(true);

	}

	@Override
	public void onConferenceSelected(long conferenceId) {
		EventsListFragment eventsListFragment = EventsListFragment.newInstance(conferenceId, getNumColumns());
		showEventsFragment(eventsListFragment);
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);
	}

	private void showBookmarksFragment() {
		EventsListFragment bookmarksFragment = EventsListFragment.newInstance(EventsListFragment.BOOKMARKS_LIST_ID, getNumColumns());
		showEventsFragment(bookmarksFragment);
	}

	private void showInProgressFragment() {
		EventsListFragment progressEventsFragment = EventsListFragment.newInstance(EventsListFragment.IN_PROGRESS_LIST_ID, getNumColumns());
		showEventsFragment(progressEventsFragment);
	}

	private void showEventsFragment(EventsListFragment eventsListFragment) {
		FragmentManager fm = getSupportFragmentManager();
		Fragment oldFragment = fm.findFragmentById(R.id.fragment_container);

		TransitionInflater transitionInflater = TransitionInflater.from(this);
		oldFragment.setExitTransition(
				transitionInflater.inflateTransition(android.R.transition.fade));
		eventsListFragment.setEnterTransition(
				transitionInflater.inflateTransition(android.R.transition.slide_right));

		Slide slideTransition = new Slide(Gravity.RIGHT);
		eventsListFragment.setEnterTransition(slideTransition);

		FragmentTransaction ft = fm.beginTransaction();
		ft.replace(R.id.fragment_container, eventsListFragment);
		ft.setReorderingAllowed(true);
		ft.addToBackStack(null);
		ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
		ft.commit();
		resetToolbar();
	}

	@Override
	public void onEventSelected(PersistentEvent event, View v) {
		EventDetailsActivity.Companion.launch(this, event.getEventId());
//		EventDetailsFragment detailsFragment = EventDetailsFragment.Companion.newInstance(event.getEventId());
//		FragmentManager fm = getSupportFragmentManager();
//
//		detailsFragment.setAllowEnterTransitionOverlap(true);
//		detailsFragment.setAllowReturnTransitionOverlap(true);
//
//		FragmentTransaction ft = fm.beginTransaction();
//		ft.replace(R.id.fragment_container, detailsFragment);
//		ft.addToBackStack(null);
//		ft.setReorderingAllowed(true);
//
//		View thumb = v.findViewById(R.id.imageView);
//		ft.addSharedElement(thumb, ViewCompat.getTransitionName(thumb));
//
//		ft.commit();
	}

	@Override
	public void setToolbarTitle(String title) {
		toolbar.setTitle(title);
	}

	private void showAboutPage() {
		Intent intent = new Intent(this, AboutActivity.class);
		startActivity(intent);
	}

	@Override
	public void onBackPressed() {
		resetToolbar();
		super.onBackPressed();
	}
}

File: touch/src/main/java/de/nicidienase/chaosflix/touch/browse/BrowseActivity.kt
package de.nicidienase.chaosflix.touch.browse

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.os.PersistableBundle
import android.transition.TransitionInflater
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import de.nicidienase.chaosflix.common.mediadata.entities.recording.persistence.Conference
import de.nicidienase.chaosflix.common.mediadata.entities.recording.persistence.Event
import de.nicidienase.chaosflix.common.mediadata.entities.streaming.StreamUrl
import de.nicidienase.chaosflix.common.viewmodel.BrowseViewModel
import de.nicidienase.chaosflix.common.viewmodel.ViewModelFactory
import de.nicidienase.chaosflix.touch.NavigationActivity
import de.nicidienase.chaosflix.touch.OnEventSelectedListener
import de.nicidienase.chaosflix.touch.R
import de.nicidienase.chaosflix.touch.about.AboutActivity
import de.nicidienase.chaosflix.touch.browse.cast.CastService
import de.nicidienase.chaosflix.touch.browse.download.DownloadsListFragment
import de.nicidienase.chaosflix.touch.browse.eventslist.EventsListActivity
import de.nicidienase.chaosflix.touch.browse.eventslist.EventsListFragment
import de.nicidienase.chaosflix.touch.browse.streaming.LivestreamListFragment
import de.nicidienase.chaosflix.touch.browse.streaming.StreamingItem
import de.nicidienase.chaosflix.touch.databinding.ActivityBrowseBinding
import de.nicidienase.chaosflix.touch.eventdetails.EventDetailsActivity
import de.nicidienase.chaosflix.touch.playback.PlayerActivity
import de.nicidienase.chaosflix.touch.settings.SettingsActivity

class BrowseActivity : AppCompatActivity(),
        ConferencesTabBrowseFragment.OnInteractionListener,
        LivestreamListFragment.InteractionListener,
        OnEventSelectedListener {

    private var drawerOpen: Boolean = false

    private val TAG = BrowseActivity::class.simpleName

    private lateinit var drawerToggle: ActionBarDrawerToggle
    private lateinit var binding: ActivityBrowseBinding

    private lateinit var viewModel: BrowseViewModel

    protected val numColumns: Int
        get() = resources.getInteger(R.integer.num_columns)

    private lateinit var castService: CastService

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_browse)

        castService = CastService(this)

        val navigationView = findViewById<NavigationView>(R.id.navigation_view)
        navigationView.setNavigationItemSelectedListener { item ->
            when (item.itemId) {
                R.id.nav_recordings -> showConferencesFragment()
                R.id.nav_bookmarks -> showBookmarksFragment()
                R.id.nav_inprogress -> showInProgressFragment()
                R.id.nav_about -> showAboutPage()
                R.id.nav_streams -> showStreamsFragment()
                R.id.nav_downloads -> showDownloadsFragment()
                R.id.nav_preferences -> showSettingsPage()
                R.id.nav_new_navigation -> {
                    startActivity(Intent(this, NavigationActivity::class.java))
                    finish()
                }
                else -> Snackbar.make(binding.drawerLayout, "Not implemented yet", Snackbar.LENGTH_SHORT).show()
            }
            binding.drawerLayout.closeDrawers()
            true
        }

        if (savedInstanceState == null) {
            showConferencesFragment()
        }
        viewModel = ViewModelProviders.of(this, ViewModelFactory.getInstance(this)).get(BrowseViewModel::class.java)

        viewModel.state.observe(this, Observer { event ->
            if (event == null) {
                return@Observer
            }
            val errorMessage = event.error
            if (errorMessage != null) {
                Snackbar.make(binding.root, errorMessage, Snackbar.LENGTH_SHORT).show()
            }
            when (event.state) {
                BrowseViewModel.State.ShowEventDetails -> {
                    event.data?.let {
                        EventDetailsActivity.launch(this, it)
                    }
                }
            }
        })
    }

    fun setupDrawerToggle(toolbar: Toolbar?) {
        if (toolbar != null) {
            drawerToggle = object : ActionBarDrawerToggle(this, binding.drawerLayout,
                    toolbar, R.string.drawer_open, R.string.drawer_close) {
                override fun onDrawerOpened(drawerView: View) {
                    super.onDrawerOpened(drawerView)
                    drawerOpen = true
                }

                override fun onDrawerClosed(drawerView: View) {
                    super.onDrawerClosed(drawerView)
                    drawerOpen = false
                }
            }
        } else {
            drawerToggle = object : ActionBarDrawerToggle(this, binding.drawerLayout,
                    R.string.drawer_open, R.string.drawer_close) {
                override fun onDrawerOpened(drawerView: View) {
                    super.onDrawerOpened(drawerView)
                    drawerOpen = true
                }

                override fun onDrawerClosed(drawerView: View) {
                    super.onDrawerClosed(drawerView)
                    drawerOpen = false
                }
            }
        }
        binding.drawerLayout.addDrawerListener(drawerToggle)
        drawerToggle.syncState()
    }

    override fun onPostCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onPostCreate(savedInstanceState, persistentState)
        drawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle.onConfigurationChanged(newConfig)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menu?.let {
            castService.addMediaRouteMenuItem(it)
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (drawerToggle.onOptionsItemSelected(item)) {
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onConferenceSelected(conference: Conference) {
        EventsListActivity.start(this, conference)
    }

    override fun onStreamSelected(streamingItem: StreamingItem) {
        val entries = HashMap<String, StreamUrl>()

        if (castService.connected) {
            val hdStreams = streamingItem.room.streams // .filter { it.slug.startsWith("hd-") }
            Log.i(TAG, "found ${hdStreams.size} suitable streams, starting selection")
            if (hdStreams.size > 1) {
                val dialog = AlertDialog.Builder(this)
                        .setTitle(getString(R.string.select_stream))
                        .setItems(hdStreams.map { it.display }.toTypedArray()) { _, i ->
                            val stream = hdStreams[i]
                            val keys = stream.urls.keys.toTypedArray()
                            val dialog = AlertDialog.Builder(this)
                                    .setTitle(this.getString(R.string.select_stream))
                                    .setItems(keys) { _: DialogInterface?, which: Int ->
                                        val streamUrl = stream.urls[keys[which]]
                                        if (streamUrl != null) {
                                            castService.castStream(streamingItem, streamUrl, keys[which])
                                        } else {
                                            Snackbar.make(binding.root, "could not play stream", Snackbar.LENGTH_SHORT).show()
                                        }
                                    }
                                    .create()
                            dialog.show()
                        }
                        .create()
                dialog.show()
            } else {
                Log.i(TAG, "Found no HD-Stream")
            }
        } else {
            val dashStreams = streamingItem.room.streams.filter { it.slug == "dash-native" }
            if (dashStreams.size > 0 &&
                    viewModel.getAutoselectStream()) {
                playStream(streamingItem.conference.conference,
                        streamingItem.room.display,
                        dashStreams.first().urls["dash"]
                )
            } else {
                streamingItem.room.streams.flatMap { stream ->
                    stream.urls.map { entry ->
                        entries.put(stream.slug + " " + entry.key, entry.value)
                    }
                }

                val builder = AlertDialog.Builder(this)
                val strings = entries.keys.sorted().toTypedArray()
                builder.setTitle(getString(R.string.select_stream))
                        .setItems(strings) { _, i ->
                            Toast.makeText(this, strings[i], Toast.LENGTH_LONG).show()
                            playStream(
                                    streamingItem.conference.conference,
                                    streamingItem.room.display,
                                    entries[strings[i]])
                        }
                builder.create().show()
            }
        }
    }

    private fun playStream(conference: String, room: String, streamUrl: StreamUrl?) {
        if (streamUrl != null) {
            PlayerActivity.launch(this, conference, room, streamUrl)
        }
    }

    private fun showConferencesFragment() {
        showFragment(ConferencesTabBrowseFragment.newInstance(numColumns), "conferences")
    }

    private fun showBookmarksFragment() {
        val bookmarksFragment = EventsListFragment.newInstance(EventsListFragment.TYPE_BOOKMARKS, null, numColumns)
        showFragment(bookmarksFragment, "bookmarks")
    }

    private fun showInProgressFragment() {
        val progressEventsFragment = EventsListFragment.newInstance(EventsListFragment.TYPE_IN_PROGRESS, null, numColumns)
        showFragment(progressEventsFragment, "in_progress")
    }

    private fun showStreamsFragment() {
        val fragment = LivestreamListFragment.newInstance(numColumns)
        showFragment(fragment, "streams")
    }

    private fun showDownloadsFragment() {
        val fragment = DownloadsListFragment.getInstance(numColumns)
        showFragment(fragment, "downloads")
    }

    private fun showSettingsPage() {
        val intent = Intent(this, SettingsActivity::class.java)
        startActivity(intent)
    }

    private fun showAboutPage() {
        val intent = Intent(this, AboutActivity::class.java)
        startActivity(intent)
    }

    override fun onBackPressed() {
        if (drawerOpen) {
            binding.drawerLayout.closeDrawers()
        } else {
            super.onBackPressed()
        }
    }

    protected fun showFragment(fragment: androidx.fragment.app.Fragment, tag: String) {
        val fm = supportFragmentManager
        val oldFragment = fm.findFragmentById(R.id.fragment_container)

        val transitionInflater = TransitionInflater.from(this)
        if (oldFragment != null) {
            if (oldFragment.tag.equals(tag)) {
                return
            }
            oldFragment.exitTransition = transitionInflater.inflateTransition(android.R.transition.fade)
        }
        fragment.enterTransition = transitionInflater.inflateTransition(android.R.transition.fade)

//        val slideTransition = Slide(Gravity.RIGHT)
//        fragment.enterTransition = slideTransition

        val ft = fm.beginTransaction()
        ft.replace(R.id.fragment_container, fragment, tag)
        ft.setReorderingAllowed(true)
        ft.setTransition(androidx.fragment.app.FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
        ft.commit()
    }

    override fun onEventSelected(event: Event) {
        EventDetailsActivity.launch(this, event)
    }

    companion object {
        fun launch(context: Context) {
            context.startActivity(Intent(context, BrowseActivity::class.java))
        }
    }
}
